### PR TITLE
fix(modal): update backdrop to close on mousedown/touchstart

### LINKED ISSUE #1950: Modals incorrectly dismissed by mouse release
### Bug description:
When you have a modal dialog that includes text which is expected to be copied with the mouse, users (inc. me) often somewhat imprecisely select the text and end up with the mouse outside of the modal dialog. They then release the mouse button used to select the text.

With ng-bootstrap modals this results in the unexpected behaviour of the whole modal gets dismissed.

I think the ideal solution would probably be to not listen for (click) in NgbModalWindow, and instead listen for (mousedown) events.
https://github.com/ng-bootstrap/ng-bootstrap/blob/0c03671c52500f13387deeaf31f815372d364c26/src/modal/modal-window.ts#L23

As a work-around I can disable the background click detection (option 'backdrop'), but it would be nice to still support that.

### Link to minimally-working plunker that reproduces the issue:

You can see the issue with [modal](https://ng-bootstrap.github.io/#/components/modal/examples)

Launch the first modal, then select all of the text ('One fine body…') and end with the mouse just outside the modal before you release the mouse button.

### Version of Angular, ng-bootstrap, and Bootstrap:

Angular: 4.4.6

ng-bootstrap: 1.0.0-beta.5

Bootstrap: 4.0.0-beta.2

Browsers:
 Microsoft Edge 40.15063.674.0 (Win)
 Microsoft IE 11.674 (Win)
 Chrome Version 61.0.3163.100 (Win)
 Chrome  61.0.3163.100 (Mac)

 Firefox 56.0 Mac does not exhibit the problem
 Safari 11.0.1 Mac does not exhibit the problem

### PR DESCRIPTION
**Pull request by lucasklaassen**
Closes #1950
Re-opens https://github.com/ng-bootstrap/ng-bootstrap/pull/2142

The click event was replaced with touchstart and mousedown. This way the modal closes only when the backdrop was clicked directly and did not start anywhere else, as described in the bug report.

### CODE REVIEW

**Comment by lucasklaassen** [comment]
@pkozlowski-opensource this definitely fixes the issue. I've tested this manually on desktop and on mobile devices using an iPhone emulator.

The change from using `(click)` to `(mousedown)` and `(touchstart)` prevents clicks from triggering the dismiss action on the modal when the user clicks inside of the modal first -> drags their mouse -> and then releases outside of the modal.

This is super helpful since people sometimes select text, inside of a modal. If they select text and drag their cursor a little bit farther than the modal itself, it will close the modal. If the user is building an object in the modal, this dismiss action can be quite costly to them and very annoying.

File: src/modal/modal-window.ts [hunk 1]
```diff
@@ -82,7 +83,9 @@ export class NgbModalWindow implements OnInit,
     } else {
       elementToFocus = body;
     }
-    elementToFocus.focus();
+    setTimeout(() => {
```
Comment (ymeine) [inline src/modal/modal-window.ts]: What is the purpose of the `setTimeout` here? It seems to make all the unit tests fail

**Comment by ymeine** [comment]
Hello,

Weirdly, the tests seem to be failing the same way as before on Travis.

Anyhow, I realized that we might not want to take this feature as is, because it changes the user experience: instead of dismissing the modal on release of the mouse pointer, it will as soon as it is pressed. This might be unwanted.

Another argument is that we are usually following the Bootstrap specs, and this PR is not matching it.

If you are still willing to fix the original issue but to comply to the Bootstrap way, you might want to apply the same technique they are using there: [bootstrap/modal.js at ceffc3a5fd7af7609150048f0651f025ef19672c · twbs/bootstrap](https://github.com/twbs/bootstrap/blob/ceffc3a5fd7af7609150048f0651f025ef19672c/js/src/modal.js).

I didn't check more than that, but as far as I understood, they applied a common technique which is to watch for mouse events and track the target elements. That way, when the release of the mouse pointer occurs (on `click`), it is possible to determine whether it corresponds to a real user action of dismissing the modal or not (by checking the initial and final targets somehow).

**Comment by maxokorokov** [comment]
Closing in favor of #3457

**Pull request closed without merging**